laurent22/joplin
# Mobile: Editor: Switch to a scrolling toolbar, allow adding/removing toolbar items

# Summary

This pull request:
- Switches the mobile app to a scrolling toolbar.
- Exposes whether a toolbar button is highlighted to accessibility tools.
- Allows configuring whether items are present in the toolbar or not.

A version of this pull request can be tried at https://personalizedrefrigerator.github.io/joplin/web-client/.

This pull request partially implements a design by @MartaLC for the mobile editor toolbar.

## Notes

- This should improve the Markdown toolbar accessibility:
    - The previous Markdown toolbar had focus-handling issues when expanding and collapsing the overflow menu. The new Markdown toolbar scrolls overflow, so should not have the same issues.
    - The new Markdown toolbar exposes button selection/activation state to screen readers.
- Currently, toolbar items can't be re-arranged (only included/excluded).
- The toolbar layout is saved as a list of commands and separators in the `editor.toolbarButtons` setting. If an empty array, the default layout is used.
    - For example, setting `editor.toolbarButtons` to `['textBold']` causes a single item to be shown in the toolbar.
    - This should allow implementing toolbar button reordering in the future, without changing the setting format.
- The `CommandService` logic used by the desktop app's Markdown toolbar is also used for the mobile app's toolbar.
- At present, if a user changes the toolbar layout, then installs a plugin that adds a toolbar button, the plugin's toolbar button won't be visible unless manually shown from the toolbar's setting screen.
   - If the user keeps the default layout, however, the button will be shown automatically after installing the plugin.


# Screenshots

## Small screen/web

![screenshot: Toolbar is shown at the bottom of the screen in light mode, with horizontal scroll](https://github.com/user-attachments/assets/0d6f7c03-03b2-413f-95ee-b7fd3d8fb589)

## Large screen/web

**Light mode**:
![screenshot: Toolbar/light mode, extends to edge of screen](https://github.com/user-attachments/assets/9cf17da7-4a22-4136-a45d-3b759beb689a)
![screenshot: Toolbar options shown](https://github.com/user-attachments/assets/095e0a35-4c66-4c64-8e7a-067684f9c870)


# Testing

## Automated tests

This pull request includes three automated tests. These tests check:
1. That the settings dialog can open.
3. That checking/unchecking items in the settings dialog adds/removes them from the toolbar.
4. That the "math" command is only available when math markup is enabled in the main app settings.

## Manual testing

This pull request has been tested manually by:
1. Opening an existing note.
2. Moving the cursor into an ordered list.
3. Verifying that the "ordered list" item has a shaded background:
5. Clicking the "ordered list" toggle button.
6. Verifying that the ordered list is removed.
7. Verifying that the list item no longer has a shaded background.
8. Pressing the search button.
9. Verifying that this opens the search dialog.
10. Pressing the "tags" button.
11. Verifying that this opens the tags dialog.
12. Clicking "cancel".
13. Clicking "bold".
14. Verifying that this adds `****` to the note.
15. Clicking "code".
16. Verifying that this adds ``` `` ``` to the note.
17. Clicking "link".
18. Verifying that this opens the link dialog.
19. Closing the link dialog.
20. Clicking "attach file".
21. Verifying that this opens the attachment dialog.
22. Attaching an image.
23. Deleting the image.
24. Clicking "settings".
25. Verifying that the settings dialog is visible.
26. Unchecking "math" and "code".
27. Verifying that the items have been removed from the toolbar.
28. Closing, then re-opening settings.
29. Clicking "restore default".
30. Clicking "cancel" in the confirmation dialog.
31. Verifying that "math" and "code" are still not visible.
32. Re-opening the settings dialog.
33. Clicking "restore default".
34. Clicking "ok".
35. Verifying that the code and math buttons are restored.

The above testing was done with the web client (Chrome). More limited manual testing has been done on Android 13 and, with an earlier commit, iOS 18.
